---
title: "TCGA PRAD"
author: "Peter Chernek and Levi Waldron"
date: "December 23, 2015"
output: html_document
---

# Load TCGA data

```{r, message=FALSE}
if(!require(RTCGAToolbox)){
  biocLite("devtools")
  biocLite(c("limma", "RCircos", "data.table", "RCurl", "RJSONIO"))
  biocLite("Link-NY/RTCGAToolbox")
}
library(DESeq2)
```

# Download TCGA PRAD data

```{r, cache=TRUE}
library(RTCGAToolbox)
rundates <- getFirehoseRunningDates()
analysisdates <- getFirehoseAnalyzeDates()
prad <- getFirehoseData("PRAD", runDate=rundates[1],
                      gistic2_Date=analysisdates[1], RNAseq_Gene=TRUE, 
        miRNASeq_Gene=TRUE, RNAseq2_Gene_Norm=TRUE, CNA_SNP=TRUE,
        CNV_SNP=TRUE, CNA_Seq=TRUE, CNA_CGH=TRUE,  Methylation=TRUE,
        Mutation=TRUE, mRNA_Array=TRUE, miRNA_Array=TRUE, RPPA=TRUE)
```

# Extract available data types


```{r}
choices <- tolower(gsub("_", "", c("RNAseq_Gene", "miRNASeq_Gene",
             "RNAseq2_Gene_Norm", "CNA_SNP", "CNV_SNP", "CNA_Seq",
             "CNA_CGH", "Methylation", "Mutation", "mRNA_Array",
             "miRNA_Array", "RPPA")))
```

```{r, cache=TRUE}
dses <- lapply(choices, function(choice) try(extract(prad, choice, 
                                            clinic=TRUE),
                                             silent=TRUE))
names(dses) <- choices
dses
```

```{r, cache=TRUE}
eset.rnaseq <- extract(prad, "rnaseq2genenorm")
write.csv(exprs(eset.rnaseq), file="prad_rnaseq.csv")
write.csv(pData(eset.rnaseq), file="prad_clinical.csv")
saveRDS(eset.rnaseq, file="prad_eset.rds")
```

To load the eset again:

```{r, cache=TRUE}
eset.rnaseq <- readRDS("prad_eset.rds")
```

# Make a histogram of PSA (KLK3) expression

```{r}
hist(exprs(eset.rnaseq["KLK3", ]))
hist(log(exprs(eset.rnaseq["KLK3", ])))
```

# See what clinical data are available by default:

```{r, eval=FALSE}
summary(pData(eset.rnaseq))
```

# Look for association between KLK3 expression and clinical PSA

```{r}
psadat <- data.frame(psa=as.numeric(as.character(eset.rnaseq$patient.stage_event.psa.psa_value)),
                     klk3=t(exprs(eset.rnaseq["KLK3", ])))
psadat.complete <- psadat[complete.cases(psadat), ]
plot(KLK3 ~ psa, data=psadat.complete, xlab="clinical PSA", ylab="KLK3 tumor expression", log="xy")
fit <- lowess(x=psadat.complete$psa, y=psadat.complete$KLK3)
lines(fit, col="red", lw=3)
```

## Is there an association between PSA in the urine and KLK3 in the tumor?

```{r}
cor.test(x=psadat$KLK3,  y=psadat$psa, method="spearman")
```


# Clinical psa as a function of ethnicity

Need to get the better race variable with complete data from Tiffany:
```{r}
table(eset.rnaseq$race)
boxplot(psadat$psa ~ eset.rnaseq$race, ylab="PSA")
```

# MicroRNA dataset
Create and save to disk:
```{r}
library(Biobase)
eset <- extract(prad, "miRNASeq_Gene")
saveRDS(eset, file="prad_mirna_eset.rds")
eset <- readRDS("prad_mirna_eset.rds")
```

```{r}
mergeVecs <- function(x1, x2){
  ##x1 and x2 are vectors to be merged.
  ##x1 will be over-written by x2, and in case of conflict, x2 takes priority
  if(!identical(length(x1), length(x2))) stop("x1 and x2 must have the same length")
  if(!identical(class(x1), class(x2))) stop("x1 and x2 must have the same class")
  x1[is.na(x1)] = x2[is.na(x1)]
  mismatches <- which(x1 != x2)
  if(length(mismatches) > 0){
    warning(paste("There were mismatches in positions:", paste0(mismatches, collapse=", ")))
    x1[mismatches] = x2[mismatches]
  }
  return(x1)
}
mergeVecs(x1=c(1, 2, 3, NA), x2=c(2, 2, NA, 4))
```

```{r}
eset$racevar = factor(mergeVecs(as.character(eset$patient.clinical_cqcf.race), as.character(eset$patient.race)))
eset$batch <- factor(eset$batch)
eset = eset[, eset$racevar %in% c("white", "black or african american")]
eset$gleason <- as.numeric(eset$gleason_score)
```

## DESeq2 correcting for batch and Gleason score

```{r, cache=TRUE}
countData <- exprs(eset)
colData <- pData(eset)
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ gleason + batch_number + racevar)
dds <- DESeq(dds)
```

```{r, cache=TRUE}
res.batch.gleason <- results(dds, contrast=c("racevar", "black or african american", "white"))
res.batch.gleason <- res.batch.gleason[order(res.batch.gleason$pvalue), ]
write.csv(res.batch.gleason, file="DESeq2_batch.gleason.csv")
```

```{r}
(res.batch.gleason.sig <- res.batch.gleason[which(res.batch.gleason$padj < 0.05), ])
```

### Heatmap

```{r heatmapbatchgleason}
library(pheatmap)
nt <- normTransform(dds) # defaults to log2(x+1)
log2.norm.counts <- assay(nt)[rownames(countData) %in% rownames(res.batch.gleason.sig), ]
df <- as.data.frame(colData(dds)[, c("gleason","racevar")])
##countData <- countData #c("hsa-mir-1-2","hsa-mir-1226","hsa-mir-133a-1","hsa-mir-133a-2", "hsa-mir-133b", "hsa-mir-137", "hsa-mir-592"), ]
pheatmap(log2.norm.counts, cluster_rows=TRUE, show_colnames=FALSE, cluster_cols=TRUE,
         clustering_distance_rows="correlation", 
         clustering_distance_cols="correlation", 
         scale="none", annotation_col=df)
```

Boxplot of the top differentially expressed miRNA
```{r boxplot}
topmir=rownames(res.batch.gleason.sig)[1]
boxplot(log2.norm.counts[topmir, ] ~ df$racevar,
        xlab="Race", ylab="log2 expression", main=topmir)
stripchart(log2.norm.counts[topmir, ] ~ df$racevar, vertical=TRUE, method="jitter",
            pch = 21, col = "maroon", bg = "bisque", 
            add = TRUE)
```

## DESeq2 correcting for batch only

```{r deseqbatchonly, cache=TRUE}
dds2 <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ batch_number + racevar)
dds2 <- DESeq(dds2)
```

```{r resbatchonly, cache=TRUE}
res.batch <- results(dds2, contrast=c("racevar", "black or african american", "white"))
res.batch <- res.batch[order(res.batch$pvalue), ]
write.csv(res.batch, file="DESeq2_batch.csv")
```

```{r}
(res.batch.sig <- res.batch[which(res.batch$padj < 0.05), ])
```

### Heatmap

```{r heatmapbatchonly, echo=FALSE}
library(pheatmap)
nt <- normTransform(dds2) # defaults to log2(x+1)
log2.norm.counts <- assay(nt)[rownames(countData) %in% rownames(res.batch.sig), ]
df <- as.data.frame(colData(dds2)[, c("gleason","racevar")])
##countData <- countData #c("hsa-mir-1-2","hsa-mir-1226","hsa-mir-133a-1","hsa-mir-133a-2", "hsa-mir-133b", "hsa-mir-137", "hsa-mir-592"), ]
pheatmap(log2.norm.counts, cluster_rows=TRUE, show_colnames=FALSE, cluster_cols=TRUE,
         clustering_distance_rows="correlation", 
         clustering_distance_cols="correlation", 
         scale="none", annotation_col=df)
```

# Not needed, here for posterity. Loading & checking Tiffany's racevar.
Load Tiffany's file with the race variable and make its barcodes equivalent to eset:
```{r}
racevar <- read.csv("racevariable.csv", stringsAsFactors = FALSE)
racevar[, 1] <- paste0(racevar[, 1], "-01")
racevar[, 1] <- gsub(".", "-", racevar[, 1], fixed=TRUE)
```

For interests' sake, which patients are in one dataset but not the other?

```{r}
racevar[, 1][!racevar[, 1] %in% sampleNames(eset)]
sampleNames(eset)[!sampleNames(eset) %in% racevar[, 1]]
```

Keep only patients in both eset and racevar, and match up the rows:
```{r}
eset2 <- eset[, sampleNames(eset) %in% racevar[, 1]]
racevar <- racevar[racevar[, 1] %in% sampleNames(eset2), ]
racevar <- racevar[match(sampleNames(eset2), racevar[, 1]), ]
all.equal(sampleNames(eset2), racevar[, 1])
```

Add racevar[, 2] to eset2:
```{r}
eset2$racevar <- racevar[, 2]
summary(eset2$racevar == eset2$race)
```